lib/Dialect/ESI/runtime/python/esiaccel/cosim/simulator.py

```diff
@@ -198,6 +259,54 @@ def run_proc(self, gui: bool = False) -> SimProcess:
       time.sleep(0.05)
     return SimProcess(proc=simProc, port=port)
 
+  def _start_process_with_callbacks(self, cmd: List[str],
+                                    env: Optional[Dict[str, str]],
+                                    cwd: Optional[Path],
+                                    stdout_cb: Optional[Callable[[str], None]],
+                                    stderr_cb: Optional[Callable[[str], None]],
+                                    wait: bool) -> subprocess.Popen | int:
+    """Start a subprocess and stream its stdout/stderr to callbacks.
+
+    If wait is True, blocks until process completes and returns its exit code.
+    If wait is False, returns the Popen object (threads keep streaming).
+    """
+    proc = subprocess.Popen(cmd,
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE,
+                            env=env,
+                            cwd=cwd,
+                            text=True,
+                            preexec_fn=os.setsid)
+
+    def _reader(pipe, cb):
+      if pipe is None:
+        return
+      for raw in pipe:
+        if raw.endswith('\n'):
+          raw = raw[:-1]
+        if cb:
+          try:
+            cb(raw)
+          except Exception:
+            # Swallow callback exceptions to avoid killing streaming threads.

```

Comment:
I don't think it's ever a good idea to just swallow exceptions. Maybe output it via the callback?

Reply:
I don't think it's a good idea to do that, when the callback has just demonstrated that it's capable of exceptional behavior. We could add a regular print, or go nuclear and `sys.exit()`... regular print for now!

Reply:
Yeah, that works too.